RFC #1042: Planning the implementation of static evaluation (#1649)
opentofu/opentofu: docs/static-evaluation.md

Implementing this initial concept will allow many of the [solutions](#solutions) below to be fully implemented.  However, there are some limitations due to module expansion which are worth considering.

## Additional Complexity due to Module Expansion

The concepts of `for_each` and `count` were grafted on to the codebase in a way that has added significant complexity and limitations. When a module block contains a `for_each` or `count` all of the nodes (resources/variables/locals/etc...) will be created multiple times, one copy per "instance".

One common example of a limitation would be to use different providers for different module instances:
```hcl
# main.tf
module "mod" {
        for_each = {"us" = "first", "eu" = "second"}
        source = "./mod"
        name = each.key
        providers {
          aws = provider.aws[each.value]
        }
}
```

As the provider requirements are baked into the module itself, the multiple "instances" don't have any concept of providers per instance. This becomes even more complex when you consider that these providers might be passed through a complex tree of modules before they are directly used.

> There are two potential solutions detailed below in [Provider Iteration](#provider-iteration), each with their own tradeoffs.  Depending on the decision made there, we may not allow each.key/vaue and count to be used in module sources (at least for now).

Reviewer: Resolved in [8f763bc](https://github.com/opentofu/opentofu/pull/1649/commits/8f763bc102050bada4e72676c69ea7ba6cd558ce)

## Plan of Attack

Between the initial implementation, the simple solutions, and the provider complexity, we are talking about a significant amount of work likely spread across multiple releases.

We can not take the approach of hacking on a feature branch for months or freezing all related code. It's unrealistic and unfair to other developers.

Instead, we can break this work into smaller discrete and testable components, some of which may be easy to work on in parallel.

> If we design an interface for the static evaluator and wire a noop implementation through the config package, work on all of the major solutions can be started. Those solutions will then become functional as pieces of the static evaluator are implemented.

Reviewer: Resolved in [8f763bc](https://github.com/opentofu/opentofu/pull/1649/commits/8f763bc102050bada4e72676c69ea7ba6cd558ce)

Example:

main.tf
```hcl
variable "input_value" {}
locals {
  hash = md5sum(var.input_value)
}
module "mod" {
  source = "./some-source"
  value = local.hash
}
```

pseudocode
```go
buildConfig(path = ".", ctx = StaticContextFromTFVars(command.TFVars))
  config = Config{}

  config.Module = loadModule(path, ctx)
    module = localModuleFiles(".")
    ctx.AddVariables(module.Variables)
    ctx.AddLocals(module.Locals)
    return module

  for call in config.Module.ModuleCalls {
    args = ctx.Evaluate(call.Arguments)
    subCtx := ctx.SubContext(call.Source, args)
    buildConfig(call.Source, subCtx)
  }
```

### Static Context Design
At the heart of the project lies an evaluation context, similar to what currently exist in the tofu and lang package. It must serve a similar purpose, but has some differing requirements.

Any static evaluator must be able to:
* Evaluate a hcl expression or block into a single cty value
  - Provide detailed insight into why a given expression or block can not be turned into a cty value

> * Be constructed with variables derived from a parent static context

Reviewer: Resolved in [8f763bc](https://github.com/opentofu/opentofu/pull/1649/commits/8f763bc102050bada4e72676c69ea7ba6cd558ce)

* Re-use existing components of the tofu and lang packages with new plumbing
  - Can build on top of existing tested logic
  - Somewhat flexible as components can be swapped out as needed
  - May require refactoring existing componets we wish to use
  - May accidentally break other packages due to poor existing testing
* Re-use current evaluator/scope constructs in tofu and lang packages
  - Would require re-designing these components to function in either mode
  - Would come with most of the core logic already implemented
  - High likelyhood of breaking other packages due to poor existing testing
  - Would likely require some ergonomic gymnastics depending on scale of refactoring

This will need to be investigated and roughly prototyped, but all solutions should fit a similar enough interface to not block development of dependent tasks. We should design the interface first, based on the requirements of the initial prototype.

## Solutions
### Module Sources
Module sources must be known at init time as they are downloaded and collated into .terraform/modules. This can be implemented by inspecting the source hcl.Expression using the static evaluator scoped to the current module.

> This is relatively straight forward once the core is implemented, but will require some more in-depth changes to support for_each/count later on.

Reviewer 2: nvm I see that we do have such considerations in the `static-module-expansion.md`

Without module pre-expansion / support for for_each/count, the process would look like:
* Create a SourceExpression field in config.ModuleCall and don't set the "config.ModuleCall.Source" field initially
* Use the static context available during NewModule constructor to evaluate all of the config.ModuleCall source fields and check for bad references and other errors.

### Provider Iteration

In [#300](https://github.com/opentofu/opentofu/issues/300), users describe how supporting for_each for configuring providers will allow much DRYer and simpler configurations.

> [!Note]
> Please read [Provider References](provider-references.md) before diving into this section!

#### Proposed Changes

> The first change is to support static variables in the provider config block.  This can then be extended to support for_each/count and be expanded at the end of the `config.NewModule()` function, similar to how module.ProviderLocalNames is generated.  This piece is fairly straightforward and can be done relatively easily.

Reviewer 3: Can you please provide an example for this case? I think I got it, but it'll reassure me if I'll see the TF code describing this scenario.

* Provider mappings for modules and resources are hard-coded at config load time on an *unexpanded* view of the config/graph structures
* Provider configurations are [pruned during graph processing](./provider-references.md#Provider-Workflow)

Potential implementation paths are explored in [Static Evaluation of Providers](./static-evaluation-providers.md)

#### Questions

Should variables be allowed in required_providers now or in the future?  Could help with versioning / swapping out for testing?

In #300, there's also a discussion on allowing variable names in provider aliases.
Example:
```
# Why would you want to do this?  It looks like terraform deprecated this some time after 0.11.
provider "aws" {
  alias = var.foo
}
```

### Backend Configuration

Once the core is implemented, this is probably the easiest solution to implement.

Notes from initial prototyping:
* The configs.Backend saves the config body during the config load and does not evaluate it
* backendInitFromConfig() in command/meta_backend.go is what evaluates the body
  - This happens before the graph is constructed / evaluated and can be considered an extension of the config loading stage.

> * We can stash a copy of the StaticContext in the configs.Backend and use it in backendInitFromConfig() to provide do the decoding.

Reviewer 3: ```suggestion
* We can stash a copy of the StaticContext in the configs.Backend and use it in backendInitFromConfig() do the decoding.
```

Reviewer 3: Also, in "to do the decoding", you mean to evaluate the references instead, right?

  - There are a few ways to do this, stashing it there was a simple way to get it working in the prototype.
* Don't forget to update the configs.Backend.Hash() function as that's used to detect any changes

### Lifecycle Attributes

Not yet investigated in depth.

### Variable defaults/validation?

Not sure we are doing this one at this juncture. It may have been passed on before due to complexity around providers, or simply that coalesce + locals exists.

### Provisioners

Not yet investigated in depth.

### Moved blocks

Not yet investigated in depth.

## Encryption

Not yet investigated in depth.

## Blockers:
### Testing

Existing testing within OpenTofu is fragmented and more sparse than we would like. Additional test coverage will be needed before, during and after many of the following items.

Code coverage should be inspected before refactoring of a component is undertaken to guide the additional test coverage required. We are not aiming for 100%, but should use it as a tool to understand our current testing.

A comprehensive guide on e2e testing should be written, see #1536.

## Unknowns:
### Providers variables

> Providers may have configuration that depends on variables and dynamic values, such as resources from other providers. There is a odd workaround within the internal/tofu package where variable values may be requested during the graph building phase. This is an odd hack and may need to be reworked for the providers iteration above.

Reviewer 3: Can you add a link to where it exists in the code? Also, a link to the documentation discussing it if it exists.

### Core functions
Do we want to support the core OpenTofu functions in the static evaluation context? Probably as it would be fairly trivial to hook in.
### Provider functions
Do we want to support provider functions during this static evaluation phase? I suspect not, without a good reason as the development costs may be significant with minimal benefit. It is trivial to detect someone attempting to use a provider function in an expression/body and to mark the expression result as dynamic.
### Module Expansion Disk Copy

> As described in #issue, large projects may incur a large cost to build a directory for every remote module.  *If* we are expanding modules in a static context when possible, that implies that we will also be building a directory for every remote module instance.

Reviewer: #issue?

TODO example for this, as well as an example error message.
TODO the whole previous paragraph should be rewritten to describe user experience, not technical details.

Implementing this initial concept will allow many of the [solutions](#solutions) below to be fully implemented.  However, there are some limitations due to module expansion which are worth considering.

## Additional Complexity due to Module Expansion

The concepts of `for_each` and `count` were grafted on to the codebase in a way that has added significant complexity and limitations. When a module block contains a `for_each` or `count` all of the nodes (resources/variables/locals/etc...) will be created multiple times, one copy per "instance".

One common example of a limitation would be to use different providers for different module instances, which does not work today:

```hcl
# main.tf
module "mod" {
        for_each = {"us" = "first", "eu" = "second"}
        source = "./mod"
        name = each.key
        providers {
          aws = provider.aws[each.value] # Currently only hard-coded values are supported here
        }
}

> ```

Author: Yes

As the provider requirements are baked into the module itself, the multiple "instances" don't have any concept of providers per instance. This becomes even more complex when you consider that these providers might be passed through a complex tree of modules before they are directly used.

There are two potential solutions detailed below in [Provider Iteration](#provider-iteration), each with their own tradeoffs.

A related problem is using `each.key/value` or `count` in module sources.  The sources don't have any concept of "module instances" from `for_each/count` and have similar limitations to provider iteration.  This a much lower priority item that is likely not going to be considered for the foreseeable future. TODO reference module expansion document.

## Plan of Attack

> Between the initial implementation, the simple Solutions, and the provider complexity, we are talking about a significant amount of work likely spread across multiple releases.

Reviewer 4: If this feature will have multiple releases, can a structured plan be created to detail which features will be rolled out in each release?

Author: That's a good question.  I think I need to re-structure this RFC based on the conversations that have been happening around the new RFC process PR.  I think it is actually going to be two different RFCs in the end to properly scope the work.

> [!NOTE]
> A common misconception is that modules are "objects". However, modules more closely resemble "namespaces" and can cross reference each other's vars/outputs as long as there is no reference loop.

## Initial implementation

As you can see above, the lack of a building and managing evaluation contexts during the config loading stage prevents any expressions with references from being evaluated. Only primitive types and expressions are allowed during that stage.

By introducing the ability to build and manage evaluation contexts during config loading, we would open up the ability for *certain* references to be evaluated during the config loading process.

For example, many users expect to be able to use `local` values within `module -> source` to simplify upgrades and DRY up their configuration. This is not currently possible as the value of `module -> source` *must* be known during the config loading stage and can not be deferred until graph evaluation.

> By utilizing Traversals/References, we can track what values are statically known throughout the config loading process. This will follow a similar pattern to the graph reference evaluation (with limitations) and may or may not re-use much of it's code.

Author: They are explained above, now with links.

We can not take the approach of hacking on a feature branch for months or freezing all related code. It's unrealistic and unfair to other developers.

Instead, we can break this work into smaller discrete and testable components, some of which may be easy to work on in parallel.

If we design an interface for the static evaluator and wire a noop implementation through the config package, work on all of the major solutions can be started. Those solutions will then become functional as pieces of the static evaluator are implemented.

Additionally, the module iteration work can likely be broken into stages, implementing changes package by package and keeping around an address translation layer (detailed below).

With this piece by piece approach, we can also add testing before, during, and after each component is added/modified.

The OpenTofu core team should be the ones to do the majority of the core implementation and the module expansion work.  If community members are interested, many of the solutions are isolated and well defined enough for them to be worked on independently of the core team.

> ## Progress Overview:

Author: This will be removed as this is translated to the new RFC layout

Example:

main.tf
```hcl
variable "input_value" {}
locals {
  hash = md5sum(var.input_value)
}
module "mod" {
  source = "./some-source"
  value = local.hash
}
```

pseudocode
```go
buildConfig(path = ".", ctx = StaticContextFromTFVars(command.TFVars))
  config = Config{}

  config.Module = loadModule(path, ctx)
    module = localModuleFiles(".")
    ctx.AddVariables(module.Variables)
    ctx.AddLocals(module.Locals)
    return module

  for call in config.Module.ModuleCalls {
    args = ctx.Evaluate(call.Arguments)
    subCtx := ctx.SubContext(call.Source, args)
    buildConfig(call.Source, subCtx)
  }
```

### Static Context Design
At the heart of the project lies an evaluation context, similar to what currently exist in the tofu and lang package. It must serve a similar purpose, but has some differing requirements.

Any static evaluator must be able to:
* Evaluate a hcl expression or block into a single cty value
  - Provide detailed insight into why a given expression or block can not be turned into a cty value

> * Be scoped to a given path

Reviewer 5: Can you expand on what this means? is it around child modules?

## Graph Reference Evaluation

After the config is fully loaded, it is transformed and processed into nodes in a graph. These nodes use the "OpenTofu References" present in their blocks/attributes (the ones not evaluated in config loading) to build both the dependency edges in the graph, and eventually an evaluation context once those references are available.

This theoretically simple process is deeply complicated by the module dependency tree and expansion therein. The graph is dynamically modified due to `for_each` and `count` being evaluated as their required references are made available. The majority of the logic in this process exists within the `tofu` and `lang` package which are somewhat tightly coupled.

For example, a module's `for_each` statement may require data from a resource: `for_each = resource.aws_s3_bucket.foo.tags`. Before it could be evaluated, the module must wait for "OpenTofu Resource Reference aws_s3_bucket.foo" to be available. This would be represented as a dependency edge between the module node and the specific resource node. The evaluation context would then include `{"resource": {"aws_s3_bucket": {"foo": {"tags": <provided value>}}}}`.

> [!NOTE]

> > A common misconception is that modules are "objects". However, modules more closely resemble "namespaces" and can cross reference each other's vars/outputs as long as there is no reference loop.

Reviewer 6: Out of curiosity, how does tofu handle a module reference loop situation?

Author: It currently prints out two different error messages, which are not super clear:
```
╷
│ Error: Reference to undeclared resource
│ 
│   on main.tf line 4, in module "a":
│    4:         in = b.out
│ 
│ A managed resource "b" "out" has not been declared in the root module.
╵
╷
│ Error: Reference to undeclared resource
│ 
│   on main.tf line 9, in module "b":
│    9:         in = a.out
│ 
│ A managed resource "a" "out" has not been declared in the root module.
```

# Implementing Init-time static evaluation of constant variables and locals

As initially described in https://github.com/opentofu/opentofu/issues/1042, many users of OpenTofu are expect to be able to use variables and locals in a variety of locations that are currently not supported. To understand why this is, we need to peek under the hood and understand how and why OpenTofu evaluates expressions in configuration.

## Expressions

The evaluation of expressions (`1 + var.bar` for example) depends on required values and functions used in the expression. In that example, you would need to know the value of `var.bar`. That dependency is known via a concept called "HCL Traversals", which represent an attribute access path and can be turned into strongly typed "OpenTofu References". In practice, you would say "the expression depends on an OpenTofu Variable named bar".

> Once you know what the requirements are for an expression (`hcl.Expression`), you can build up an evaluation context (`hcl.EvalContext`) to provide those requirements or return an error.  In the above example, the evaluation context would include `{"var": {"bar": <somevalue>}`.

Reviewer 5: might be nice to link to the code here (godoc is fine)

> Expression evaluation is currently split up into two stages: config loading and reference evaluation

Reviewer 5: ```suggestion
Expression evaluation is currently split up into two stages: config loading and graph reference evaluation
```

## Config Loading

During configuration loading, the hcl or json config is pulled apart into Blocks and Attributes. A Block contains Attributes and nested Blocks. Attributes are simply named expressions (`foo = 1 + var.bar` for example).

These Blocks and Attributes are abstract representations of the configuration which have not yet been evaluated into actionable values. Depending on the handing of the given block/attribute, either the abstract representation is kept or it is evaluated into a real value for use.

As a concrete example, the `module -> source` field must be known during configuration loading as it is required to continue the next iteration of the loading process.  However attributes like `module -> for_each` may depend on attribute values from resources or other pieces of information not known during config loading and therefore are stored as an expression for later evaluation.

> No evaluation context is built or provided during the entire config loading process.  **Therefore, no functions, locals, or variables may be used when during config loading due to the lack of context.  This limitation is what we wish to resolve**.

Reviewer 5: ```suggestion
No evaluation context is built or provided during the entire config loading process.  **Therefore, no functions, locals, or variables may be used during config loading due to the lack of context.  This limitation is what we wish to resolve**.
```

## Graph Reference Evaluation

After the config is fully loaded, it is transformed and processed into nodes in a graph. These nodes use the "OpenTofu References" present in their blocks/attributes (the ones not evaluated in config loading) to build both the dependency edges in the graph, and eventually an evaluation context once those references are available.

> This theortically simple process is deeply complicated by the module dependency tree and expansion therein. The graph is dynamically modified due to `for_each` and `count` being evaluated as their required references are made available. The majority of the logic in this process exists within the `tofu` and `lang` package and are somewhat tightly coupled.

Reviewer 5: ```suggestion
This theoretically simple process is deeply complicated by the module dependency tree and expansion therein. The graph is dynamically modified due to `for_each` and `count` being evaluated as their required references are made available. The majority of the logic in this process exists within the `tofu` and `lang` package which are somewhat tightly coupled.
```

For example, a module's `for_each` statement may require data from a resource: `for_each = resource.aws_s3_bucket.foo.tags`. Before it could be evaluated, the module must wait for "OpenTofu Resource Reference aws_s3_bucket.foo" to be available. This would be represented as a dependency edge between the module node and the specific resource node. The evaluation context would then include `{"resource": {"aws_s3_bucket": {"foo": {"tags": <provided value>}}}}`.

> Note: A common misconception is that modules are "objects". Modules more closely resemble "namespaces" and can cross reference each other's vars/outputs as long as there is no reference loop.

Reviewer 5: ```suggestion
> [!NOTE]
> A common misconception is that modules are "objects". However, modules more closely resemble "namespaces" and can cross reference each other's vars/outputs as long as there is no reference loop.
```

## Initial implementation

> As you can see above, the lack of a building and managing evaluation contexts during the config loading stage prevents any expressions with references from being evaluation. Only primitive types and expressions are allowed during that stage.

Reviewer 5: ```suggestion
As you can see above, the lack of a building and managing evaluation contexts during the config loading stage prevents any expressions with references from being evaluated. Only primitive types and expressions are allowed during that stage.
```

When evaluating an Attribute/Block into a value, any missing reference must be properly reported in a way that the user can easily debug and understand. For example, a user may try to use a local that depends on a resource's value in a module's source. The user must then be told that the local can not be used in the module source field as it depends on a resource which is not yet available.  Variables used through the module tree must also be passed with their associated information. In practice this is fairly easy to track and has been prototyped during the exploration of [#1042](https://github.com/opentofu/opentofu/issues/1042).

Implementing this initial concept will allow many of the Solutions below to be fully implemented.  However, there are some limitations due to module expansion which are worth considering.

## Additional Complexity due to Module Expansion

The concepts of `for_each` and `count` were grafted on to the codebase in a way that has added significant complexity and limitations. When a module block contains a `for_each` or `count` all of the nodes (resources/variables/locals/etc...) will be created multiple times, one copy per "instance".

> One common example is using different providers for different module instances:

Reviewer 5: ```suggestion
One common example of a limitation would be to use different providers for different module instances:
```
